### PR TITLE
Generator voltage remote control support 

### PR DESCRIPTION
**Pull request by geofjamg**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
There is no generator voltage remote control support.

**What is the new behavior (if this is a feature change)?**
Generator voltage remote control  is now supported. CoordinatedReactiveControl extension is used for reactive power calculation.

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworks.java [hunk 1]
```diff
@@ -73,11 +74,16 @@ public void visitThreeWindingsTransformer(ThreeWindingsTransformer transformer,
...
                 public void visitGenerator(Generator generator) {
-                    if (!Objects.equals(generator.getRegulatingTerminal().getBusView().getBus().getId(),
-                            generator.getTerminal().getBusView().getBus().getId())) {
-                        throw new PowsyblException("Generator remote voltage control is not yet supported");
-                    }
                     lfBus.addGenerator(generator);
+
+                    String controlBusId = generator.getRegulatingTerminal().getBusView().getBus().getId();
+                    String connectedBusId = generator.getTerminal().getBusView().getBus().getId();
+                    if (!Objects.equals(controlBusId, connectedBusId)) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfNetworks.java]: Are we sure that we want to do that even if the voltage regulation is disabled ? In my use case, if I remove the remote voltage control from generators, there is no calculation.

File: src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java [hunk 2]
```diff
@@ -47,6 +57,67 @@ private static void createBusEquations(LfNetwork network, VariableSet variableSe
...
+        List<Double> qKeys = createReactiveKeys(sourceBuses);
+
+        for (LfBus sourceBus : sourceBuses) {
+            // deactivate reactive equation for all remote voltage source bus (where the generator is connected)
+            equationSystem.createEquation(sourceBus.getNum(), EquationType.BUS_Q).setActive(false);
+        }
+
+        // create (number of source bus) - 1 equation to link the reactive equations together
+        LfBus firstSourceBus = sourceBuses.get(0);
+        for (int i = 1; i < sourceBuses.size(); i++) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java]: I don't really understand this loop.

File: src/main/java/com/powsybl/openloadflow/equations/Equation.java [hunk 3]
```diff
@@ -100,10 +101,23 @@ void initTarget(LfNetwork network, double[] targets) {
                 targets[row] = network.getBus(num).getTargetV();
                 break;
 
+            case BUS_REMOTE_V:
+                targets[row] = network.getBus(num).getRemoteControlSourceBuses()
+                        .stream()
+                        .filter(LfBus::hasVoltageControl)
+                        .findFirst()
+                        .orElseThrow(() -> new IllegalStateException("None of the remote control source buses have voltage control on"))
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/equations/Equation.java]: Is it a real issue that all the generators with remote control are disabled ? I have this case on Litgrid network.

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 4]
```diff
@@ -61,6 +63,15 @@ public OpenLoadFlowParameters setDc(boolean dc) {
         return this;
     }
 
+    public boolean hasGeneratorVoltageRemoteControl() {
+        return generatorVoltageRemoteControl;
+    }
+
+    public OpenLoadFlowParameters setGeneratorVoltageRemoteControl(boolean generatorVoltageRemoteControl) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java]: Can we changed this parameter from the config.yml file ? It is required for me if I want to use the same tool (build with ittols packager) to validate LF results from different LF engines. Can we create a section ````openloadflow-default-parameters````. Or maybe this parameter should be located in the common loadflow API.

Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java]: Yes, this parameter should be in common loadflow parameters as many other one (distributed slack, etc).
We should open an issue and list parameters (from open load flow and hades2) that could be shared  between all implementations.
Persistence of default values already exists for common loadflow parameters so we could do it here once the generatorVoltageRemoteControl have been moved.

File: src/main/java/com/powsybl/openloadflow/equations/Equation.java (hunk 3 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/equations/Equation.java]: This (normal) case has already been handled in network read part so at this stage we should not go into this part of the code.

File: src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java [hunk 5]
```diff
@@ -47,6 +57,67 @@ private static void createBusEquations(LfNetwork network, VariableSet variableSe
         }
     }
 
+    private static void createLocalVoltageEquation(VariableSet variableSet, EquationSystem equationSystem, LfBus bus) {
+        equationSystem.createEquation(bus.getNum(), EquationType.BUS_V).addTerm(new BusVoltageEquationTerm(bus, variableSet));
+        equationSystem.createEquation(bus.getNum(), EquationType.BUS_Q).setActive(false);
+    }
+
+    private static void createRemoteVoltageEquations(LfBus bus, List<LfBus> sourceBuses, EquationSystem equationSystem, VariableSet variableSet) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java]: This equation system for remote control does not work on my use cases, from the first NR iteration. It meens that the issue is in the equation system, not in the reactive limits outer loop. I have tried to make something simpler: in my code test, I do as if the generator with remote control is directly connected at the remote bus and the source buses just have the reactive linked equations. 
````
equationSystem.createEquation(bus.getNum(), EquationType.BUS_REMOTE_V).addTerm(new BusVoltageEquationTerm(bus, variableSet)); // New PV bus.
equationSystem.createEquation(bus.getNum(), EquationType.BUS_Q).setActive(false); // Not a PQ bus anymore.

….

for (LfBus sourceBus : sourceBuses) {
            // deactivate reactive equation for all remote voltage source bus (where the generator is connected)
            // equationSystem.createEquation(sourceBus.getNum(), EquationType.BUS_Q).setActive(false);
           // Source buses are in fact PQ buses., but the sum of reactive power of source buses should be linked.  
        }
````

File: src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java (hunk 2 above)
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java]: I don't have found any real use case to test this part…

File: src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java (hunk 5 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java]: But maybe it is not the equation system you have in mind…

**Comment by annetill** [comment]
I have tested this feature on networks without voltage remote control, on networks with shared or not shared voltage remote control. The voltage plan is close to those we can calculate with Hades2 (https://github.com/rte-france/hades2-distribution). We can merge this feature or wait a little bit for the treatement of reactive limits in case of a shared control.

**Comment by geofjamg** [comment]
> I have tested this feature on networks without voltage remote control, on networks with shared or not shared voltage remote control. The voltage plan is close to those we can calculate with Hades2 (https://github.com/rte-france/hades2-distribution). We can merge this feature or wait a little bit for the treatement of reactive limits in case of a shared control.

We should merge it as there is already a lot of code and uncovered case is rare and cleanly handled by throwing an exception.

**Pull request merged**